### PR TITLE
Add what3words as a supported provider

### PR DESCRIPTION
**Pull request by vicchi**
Note - due to the semantics of the what3words _3 word address_ scheme, this provider returns the 3 word address as the `locality` element of the returned `Address`. The coordinates and bounding box are also always returned in their usual place in `Address`.

A valid API key is also required to use the service and for the unit tests, this can be supplied in a local copy of `phpunit.xml` as `<server name="WHAT3WORDS_API_KEY" value="a-valid-api-key" />`

### CODE REVIEW

File: tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php [hunk 1]
```diff
@@ -203,10 +203,11 @@ public function testGetReversedDataWithRealCoordinates()
         $this->assertCount(1, $result);
 
         $result = $result[0];
+        error_log(var_export($result, true));
         $this->assertInternalType('array', $result);
         $this->assertEquals(40.707507, $result['latitude'], '', 0.0001);
         $this->assertEquals(-74.011255, $result['longitude'], '', 0.0001);
-        $this->assertEquals('2', $result['streetNumber']);
+        $this->assertEquals('1', $result['streetNumber']);
```
Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php]: These changes is not related to this PR, right?

**Comment by Nyholm** [comment]
Good job 👍

**Comment by willdurand** [comment]
Try not to add `merge branch ...` as commit :)

**Comment by vicchi** [comment]
@willdurand Mea culpa!

File: tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php (hunk 1 above)
Comment (vicchi) [inline tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php]: No - but due to a back end data change it was stopping the unit tests from running clean so I tweaked the expected results. I always try to isolate changes to a specific branch but as this was a single line change I broke my own rules in this case.

File: tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php [hunk 2]
```diff
@@ -203,10 +203,11 @@ public function testGetReversedDataWithRealCoordinates()
         $this->assertCount(1, $result);
 
         $result = $result[0];
+        error_log(var_export($result, true));
```
Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php]: Is this debug code?

Comment (vicchi) [inline tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php]: Oh [expletive redacted]. Yes. I'll fix that now.

Comment (vicchi) [inline tests/Geocoder/Tests/Provider/GeocoderCaProviderTest.php]: Fixed

**Comment by willdurand** [comment]
@vicchi no worry, was just a quick check for cosmetic stuff. I think you will have to squash your commits before getting this PR merged (cf. https://davidwalsh.name/squash-commits-git).

Good job by the way!

**Comment by vicchi** [comment]
Good point. I'll squash this down. Probably won't be until tomorrow now as I'll be in transit home in a short while.

**Comment by willdurand** [comment]
No rush ;-)

**Comment by vicchi** [comment]
OK - that should be squashed now (dived onto the flaky conference wifi)

**Comment by vicchi** [comment]
I've just merged in PR #31's changes to my `gg-what3words-provider` branch and it's building clean in Travis ... if there's no other comments, can this be merged into `master`?

File: src/Geocoder/Provider/What3wordsProvider.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return $this->executeQuery($query);
+    }
+
+    public function reverse($latitude, $longitude)
+    {
+        if (null == $this->apiKey) {
+            throw new InvalidCredentialsException('No what3word API key provided.');
+        }
+
+        $query = sprintf(self::REVERSE_URL, $latitude, $longitude, $this->apiKey);
```
Comment (Nyholm) [inline src/Geocoder/Provider/What3wordsProvider.php]: Don't you need the apiKey when doing a reverse? I suggest adding a similar check as in the "geocode" function.

File: src/Geocoder/Provider/What3wordsProvider.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        return $this->executeQuery($query);
+    }
+
+    protected function executeQuery($query)
+    {
+        $content = $this->getAdapter()->get($query)->getBody();
+
+        if (!$data = json_decode($content, true)) {
```
Comment (Nyholm) [inline src/Geocoder/Provider/What3wordsProvider.php]: You should check if $data is null. 

```php
if (null === $data = json_decode($content, true)) {
```

File: src/Geocoder/Provider/What3wordsProvider.php (hunk 3 above)
Comment (vicchi) [inline src/Geocoder/Provider/What3wordsProvider.php]: Err. Am I missing something here?

There are checks for a supplied API key for both the `geocode` method (at line 36) and `reverse` method (at line 46) ...

File: src/Geocoder/Provider/What3wordsProvider.php (hunk 4 above)
Comment (vicchi) [inline src/Geocoder/Provider/What3wordsProvider.php]: Yes. I should! Update will be forthcoming. Mea culpa.

File: src/Geocoder/Provider/What3wordsProvider.php (hunk 3 above)
Comment (Nyholm) [inline src/Geocoder/Provider/What3wordsProvider.php]: oh sorry, My bad. 

I guess I reviewed this PR too fast.

Comment (Nyholm) [inline src/Geocoder/Provider/What3wordsProvider.php]: Mea culpa =)

**Comment by Nyholm** [comment]
Thank you

**Comment by vicchi** [comment]
Thanks @Nyholm. Much appreciated.
